lib/chkname.c

```diff
@@ -58,6 +63,21 @@ login_name_max_size(void)
 static bool
 is_valid_name(const char *name)
 {
+	if (streq(name, "")
+	 || streq(name, ".")
+	 || streq(name, "..")
+	 || strcaseeq(name, "none")
+	 || strcaseeq(name, "all")
+	 || strcaseeq(name, "except")
+	 || strprefix(name, "-")
+	 || strpbrk(name, " \"#,/:@")
```

Comment:
It's really hard to say where to stop here. The `|` also has a special meaning, but in lesser extend: If glibc's NSS compat is in place and somebody prepends the username with a `+` or `-`, then the user name is added into a blacklist to ignore further lookups in other NSS modules. The blacklist is a single string with names separated by `|`. In other words, the user `a|b` would add the users `a` and `b` to the blacklist. See https://sourceware.org/git/?p=glibc.git;a=blob;f=nss/nss_compat/compat-pwd.c;h=7da6dc99f7e8aac6d212df42cc33221bdc6e2f35;hb=HEAD#l1079

Does it affect shadow? If running with glibc and passwd: compat in nsswitch.conf, yes.

In short, it's really a very good idea to comply with POSIX user names in general ...

Reply:
And while at it, `&` and `;` (as well as `|`) are also special for Linux-PAM (namely, the pam_time module): https://github.com/linux-pam/linux-pam/blob/master/modules/pam_time/time.conf.5.xml#L64-L68

Reply:
> And while at it, `&` and `;` (as well as `|`) are also special for Linux-PAM (namely, the pam_time module): https://github.com/linux-pam/linux-pam/blob/master/modules/pam_time/time.conf.5.xml#L64-L68

In that link, I see mentioned `!` (which you didn't mention), and I don't see mentioned `;`.  Some lines above I do indeed see a mention to `;`.  And a few lines below I see `*`.

So, I'll add to the ban: `|`, `&`, `;`, `+`, `*`, `!`

And I do indeed agree that outside of POSIX, there be dragons!  Which is why I have the other PR, Little Bobby Tables.  :-)

Reply:
Oh, I copied the same link twice, sorry. And yes, the `*` in this context could be banned as well, even though `*` would be a valid match ... :D

Reply:
> And I do indeed agree that outside of POSIX, there be dragons!

As we see here, even inside POSIX we have the dragons. So we can stop wanting back a "safe" POSIX land.